modules/c++/re/source/RegexPCRE.cpp

```diff
@@ -70,244 +181,145 @@ re::Regex& re::Regex::operator=(const re::Regex& rhs)
+                          size_t& end)
 {
-    int numMatches(0);
-    int result(0);
-    int startOffset(0);
-
-    // Clear the output vector
-    mOvector.assign(mOvector.size(), 0);
-    numMatches = pcre_exec(mPCRE,                             // the compiled pattern
-                           NULL,                              // no extra data
-                           matchString.c_str() + startIndex,  // the subject string
-                           matchString.length() - startIndex, // the subject length
-                           startOffset,                       // starting offset
-                           flags,                             // options
-                           &mOvector[0],                      // output vector
-                           OVECTOR_COUNT);                    // output vector size
-
-    result = numMatches;
-
-    if (result == 0)
-    {
-        numMatches = OVECTOR_COUNT / 3;
-    }
-
-    if (result >= 0)
-    {
-        if (((mOvector[0] + startIndex) +
-             (mOvector[1] - mOvector[0])) >
-            (int)matchString.length() )
-        {
-            result = PCRE_ERROR_NOMATCH;
-        }
-    }
+    ScopedMatchData matchData(mPCRE);
+    const size_t numMatches = matchData.match(matchString, startIndex, flags);
 
-    if (result >= 0)
-    {
-        // output vector start offset = i*2
-        // output vector end   offset = i*2+1
-        // i = 0
-        int index = mOvector[0] + startIndex;
-        int subStringLength = mOvector[1] - mOvector[0];
-        int subStringCheck = index + subStringLength;
-        if (subStringCheck > (int)matchString.length())
-        {
-            throw RegexException(Ctxt(FmtX("Search: Match substring out of range (%d,%d) for string of length %d", index, subStringCheck, matchString.length())));
-        }
-        return matchString.substr(index, subStringLength);
-    }
-    else if (result == PCRE_ERROR_NOMATCH)
+    if (numMatches > 0)
     {
-        return std::string("");
+        begin = matchData.getOutputVector()[0];
+        end = matchData.getOutputVector()[1];
+        return matchData.getMatch(matchString, 0);
     }
     else
     {
-        throw RegexException
-            (Ctxt(FmtX("Error in searching for %s", matchString.c_str())));
+        begin = end = 0;
+        return "";
     }
 }
 
-void re::Regex::searchAll(const std::string& matchString,
-                          RegexMatch & v)
+void Regex::searchAll(const std::string& matchString, RegexMatch& v)
 {
-    std::string result = search(matchString, 0, 0);
+    size_t startIndex = 0;
+
+    size_t begin;
+    size_t end;
+    std::string result = search(matchString, startIndex, 0, begin, end);
 
-    int idx = 0;
-    while (result.size() != 0)
+    while (!result.empty())
     {
         v.push_back(result);
-        idx += (int)result.size() + mOvector[0];
-        result = search(matchString, idx, PCRE_NOTBOL);
+        startIndex = end;
```

Comment:
OK, easy part of the question first.  We use the terminology `[begin, end)` and `[first, last]` (the `[begin, end)` bit matches the STL's terminology).  But, the overlapping matches is a good call.  Actually, things are worse than that... take a look at the unit tests I just added.  The old functionality was not right; you can only increment `startIndex` to `begin + 1` safely (in your example, `startIndex = end` produced the wrong result already due to the overlap).  And actually, in making this change, the `testDotAllFlag()` function started failing which was because it wasn't right; the first test in there should have also found "3 4".  Now it does.

While we don't have the exact same condition in `sub()`, the PCRE implementation was still jumping a little too far.  Then `split()` wasn't handling the indexing for certain split patterns quite right either.  In both of these cases, you can't have overlapping patterns the same way you can in `searchAll()` at least.

Now thoroughly discouraged as it seems like all sorts of edge conditions were wrong, I added several more unittests to try to catch everything.  I think the PCRE implementation is good now, but see if you agree with the behavior of the unittests, any more edge conditions come to mind, and then also take a look at the STL implementation as it must have been off in its handling of some of these things too given that the unittests all passed previously for it.

This also means that the PCRE 5.0 implementation didn't correctly handle some of this either, so the good news I guess is we are fixing edge conditions in the code.  I'm certainly glad you took a look at these updates.